src/resdk/resources/fields.py

```diff
@@ -0,0 +1,433 @@
+
+    def _check_many(self, value, instance):
+        """Check that the many attribute is respected.
+
+        :raises AssertionError: the value is not list and many attribute is set.
+        """
+        from resdk.query import ResolweQuery
+
+        if self._many and value is not None:
+            if not isinstance(value, (list, ResolweQuery)):
+                raise ValueError(
+                    (
+                        f"Type of '{instance.__class__.__name__}.{self.public_name}' "
+                        f"must be list."
+                    )
+                )
+
+    def _check_allow_null(self, value):
+        """Check if the value is allowed to be None.
+
+        :raises ValueError: if the value is None and allow_null is False.
+        """
+        if not self._allow_null and value is None:
+            raise ValueError(f"Field {self.public_name} does not allow None value.")
+
+    def _checks_before_get(self, instance: "BaseResource"):
+        """Perform all checks before getting the value."""
+        self._check_exists(instance)
+
+    def _check_types(self, value, instance):
+        """Check the type of the value."""
+        if self._accepted_types is None or value is None:
+            return
+
+        for item in value if self._many else [value]:
+            if not isinstance(item, self._accepted_types):
+                types = f"{', '.join(_type.__name__ for _type in self._accepted_types)}"
+                many = "a list of " if self._many else ""
+                raise ValueError(
+                    (
+                        f"Type of '{instance.__class__.__name__}.{self.public_name}' "
+                        f"must be {many}'{types}'."
+                    )
+                )
+
+    def _checks_before_set(self, instance: "BaseResource", value):
+        """Perform all checks before setting the value."""
+        self._check_exists(instance)
+        self._check_many(value, instance)
+        self._check_writeable(instance)
+        self._check_allow_null(value)
+        self._check_types(value, instance)
+
+    def _is_lazy(self, value) -> bool:
+        """Check if the value must be lazy loaded."""
+        # from resdk.query import ResolweQuery
+        # return isinstance(value, ResolweQuery) or callable(value)
+        return callable(value)
+
+    def _lazy_load(self, value):
+        """Perform the lazy loading of the field value.
+
+        If the value is not lazy-loadable return it unchanged.
+        """
+        return value if not self._is_lazy(value) else value()
+
+    def status(self, instance: "BaseResource") -> FieldStatus:
+        """Return the field status."""
+        if not hasattr(instance, self._status_attribute_name):
+            setattr(instance, self._status_attribute_name, FieldStatus.UNSET)
+        return getattr(instance, self._status_attribute_name)
+
+    def changed(self, instance: "BaseResource"):
```

Comment:
I don't like that this method requires the parent instance. Is it different from `self._owner`?

Reply:
The descriptors itself (field instances) are class-level variables, so they have no instances associated with them. The `_owner` is actually a class they are assigned to, not the instance.

So all the methods for working with data require instance as argument.